File: bin/coyote_segmentator.pl
#!/usr/bin/perl -w
use strict;
use Getopt::Long;
use File::Basename;
use lib dirname (__FILE__);
use vcf2;
use strict;
use Data::Dumper;


#####                                                               #####
#                                                                       #
#  Take svdb merged sv-vcf annotated (VEP, maybe db of artefacts).      #
#  Create segments for coyote, perhaps filter really large ones?        #
#                                                                       #
####                                                                #####

# Get command line options
my $rand = int rand 1000000000;
my %opt = ();
GetOptions( \%opt, 'vcf=s', 'id=s', 'panel=s', 'normal', 'genes=s');

my %genes_panel;
my @panels = split(',',$opt{'panel'});
## read panels in order separated by comma. The base panel + specific. 
## Overlapping genes will have the information from the last in the list
foreach my $panel (@panels) { 
    read_panel($panel);
}
my $bedtools = read_vcf($opt{'vcf'});

annotate_genes($bedtools, $opt{"genes"},\%genes_panel);
sub read_vcf {
    my $fn = shift;
    my $vcf = vcf2->new('file'=>$fn );
    my $callers = callers($vcf->{header_str});
    my @callers = @$callers;
    my $bedtools = "tmp.bed";
    open (BED, ">>", $bedtools);
    while ( my $var = $vcf->next_var() ) {
        my $chrom = $var->{CHROM};
        my $start = $var->{POS};
        my $type = $var->{INFO}->{SVTYPE};
        next if ($type eq 'BND');
        my $end;
        if ( $var->{INFO}->{END} ) {
            $end = $var->{INFO}->{END};
        }
        else {
            $end = $start + abs($var->{INFO}->{SVLEN});
        }
        ## only in gatk/cnvkit
        my $probes;
        if ($var->{INFO}->{PROBES}) {
            $probes = $var->{INFO}->{PROBES};
        }
        else {
            $probes = 0;
        }
        my $fold;
        if ($var->{INFO}->{FOLD_CHANGE_LOG}) {
            $fold = $var->{INFO}->{FOLD_CHANGE_LOG};
        }
        ## for gatk-normals
        elsif ( $var->{INFO}->{gatkCN}) {
            $fold = $var->{INFO}->{gatkCN}/2;            
        }
        else {
            if ($type eq "DEL") {
                $fold = "DEL";
            }
            elsif ($type eq "DUP" || $type eq "INS") {
                $fold = "AMP"
            }
        }
        ## callers ##
        if ( $var->{INFO}->{set} eq "Intersection") {
            $callers = join('&',@callers);
        }
        else {
            $callers = $var->{INFO}->{set};
        }
        ## AMP OR DEL
        if ($type eq "DEL") {
            $type = "-";
        }
        elsif ($type eq "DUP" || $type eq "INS") {
            $type = "+";
        }

        ## ADD PR AND SR for manta-variants
        my ($PR,$SR) = (0,0);
        if ($callers =~ /manta/ ) {
            ($PR,$SR) = manta_variant($var,$opt{'id'});
        }
       

        print BED $chrom."\t".$start."\t".$end."\t".$probes."\t".$fold."\t".$type."\t".$callers."\t"."$PR:$SR\n";

        

    }
    return $bedtools;
}

sub genes {
    my $csq = shift;
    my %genes;
    foreach my $transcript ( @{ $csq }) {
        $genes{$transcript->{SYMBOL}} = 1;
    }
    my @genes;
    foreach my $gene (keys %genes) {
        unless ($gene eq "") {
            push @genes,$gene;
        }
        
    }
    my $genes = join(',',@genes);
    return $genes;
}

sub read_panel {
    my $fn = shift;
    open(PANEL, $fn) or die $!;

    while(<PANEL>) {
        chomp;
        my @line = split("\t");
        my $gene = $line[0];
        my $type = $line[1];
        my $question = $line[2];
        $genes_panel{$gene}{TYPE} = $type;
        $genes_panel{$gene}{QUESTION} = $question;
    }

}

sub get_panel_genes {
    my ($genes,$genes_panel) = @_;
    my @genes = split(',',$genes);
    my @intresting;
    foreach my $gene ( @genes ) {
        if ( $genes_panel{$gene} ) {
            push @intresting,$gene.":".$genes_panel{$gene}{TYPE}.":".$genes_panel{$gene}{QUESTION};
        }
    }
    my $any_interest = scalar(@intresting);
    my $intresting = join(',',@intresting);   
    return $any_interest,$intresting;
}

sub callers {
    my $header = shift;
    my @callers;
    if ($header =~ /gatk/ ) {
        push @callers,"gatk";
    }
    if ($header =~ /manta/ ) {
        push @callers,"manta";
    }
    if ($header =~ /cnvkit/ ) {
        push @callers,"cnvkit";
    }
    return \@callers;
}

sub manta_variant {
    my ($format, $id) = @_;

    my $PR; my $SR;
    foreach my $ind ( @{ $format->{GT} }) {
        if ($ind->{_sample_id} eq $id ) {
            $PR = $ind->{PR};
            $PR =~ s/,/\//;
            $SR = $ind->{SR};
            $SR =~ s/,/\//;
        }
    }
    return $PR,$SR;
}

sub annotate_genes {
    my ($bed, $genes, $genes_panel)  = @_;
    my $rnd = int rand 1000000000;
    my $tmp_infile = "input.$rnd.bed";
    my @overlap = `bedtools intersect -a $bed -b $genes -loj`;
    unlink ("tmp.bed");
    my $prev_reg;
    my @genes;
    my $prev_bed_str;

    my $filtered_bed = $opt{'id'}.".cn-segments.panel.bed";
    my $unfiltered_bed = $opt{'id'}.".cn-segments.bed";
    unlink( $filtered_bed );
    unlink( $unfiltered_bed );
    system(" touch $filtered_bed ");
    system(" touch $unfiltered_bed ");
    open (FILTERED, '>>', $filtered_bed);
    open (UNFILTERED, '>>', $unfiltered_bed);
            
    foreach my $line (@overlap) {
        chomp $line;
        my @f = split /\t/, $line;

        my $reg = "$f[0]\t$f[1]\t$f[2]\t";
        my $bed_str = "$f[0]\t$f[1]\t$f[2]\t$f[3]\t$f[4]\t$f[5]\t$f[6]\t$f[7]";
        if( $prev_reg and $reg ne $prev_reg ) {
            my $gene_conc = join(',',@genes);
            my ($match,$intresting) = get_panel_genes($gene_conc,$genes_panel);
            my @bed_str = split("\t",$prev_bed_str);
            if ($match > 0) { 
                print FILTERED join("\t",@bed_str[0..5]);
                print FILTERED "\t".$gene_conc."\t".$intresting."\t";
                print FILTERED join("\t",@bed_str[6..7]);
                if ($opt{'normal'}) {
                    print FILTERED "\tNORMAL";
                }
                print FILTERED "\n";
            }
            print UNFILTERED join("\t",@bed_str[0..5]);
            print UNFILTERED "\t".$gene_conc."\t";
            print UNFILTERED join("\t",@bed_str[6..7]);
            print UNFILTERED "\n";
            @genes = ();
        }
        push @genes, $f[-1];
        
        $prev_reg = $reg;
        $prev_bed_str = $bed_str;
    }
    ### PRINT LAST VARIANT, ugly code try to reduce when you have time
    my $gene_conc = join(',',@genes);
    my ($match,$intresting) = get_panel_genes($gene_conc,$genes_panel);
    my @bed_str = split("\t",$prev_bed_str);
    if ($match > 0) {
        
        print FILTERED join("\t",@bed_str[0..5]);
        print FILTERED "\t".$gene_conc."\t".$intresting."\t";
        print FILTERED join("\t",@bed_str[6..7]);
        if ($opt{'normal'}) {
            print FILTERED "\tNORMAL";
        }
        print FILTERED "\n";
    }
    print UNFILTERED join("\t",@bed_str[0..5]);
    print UNFILTERED "\t".$gene_conc."\t";
    print UNFILTERED join("\t",@bed_str[6..7]);
    print UNFILTERED "\n";
}

File: bin/coyote_segmentator.py
#!/usr/bin/env python3
"""Create Coyote CN segment BED files from merged SV calls."""

from argparse import ArgumentParser

import vcf2 as vcf
from vcf_pipeline_utils import leading_float


def _read_panel(path, panel_genes):
    """Read one coyote segmentator panel file into a shared dictionary."""
    with open(path) as fh:
        for line in fh:
            cols = line.rstrip("\n").split("\t")
            if len(cols) >= 3:
                panel_genes[cols[0]] = {"TYPE": cols[1], "QUESTION": cols[2]}


def _panel_gene_hits(genes, panel_genes):
    """Return matching panel gene annotations."""
    interesting = []
    for gene in genes.split(","):
        if gene in panel_genes:
            interesting.append(f"{gene}:{panel_genes[gene]['TYPE']}:{panel_genes[gene]['QUESTION']}")
    return len(interesting), ",".join(interesting)


def _header_callers(header_text):
    """Infer caller names from a merged SV VCF header."""
    callers = []
    for caller in ("gatk", "manta", "cnvkit"):
        if caller in header_text:
            callers.append(caller)
    return callers


def _manta_pr_sr(var, sample_id):
    """Return Manta PR and SR strings for one sample."""
    for gt in var["GT"]:
        if gt["_sample_id"] == sample_id:
            return gt.get("PR", "").replace(",", "/"), gt.get("SR", "").replace(",", "/")
    return 0, 0


def write_coyote_segment_intervals(vcf_file, sample_id, out_file):
    """Write unannotated Coyote CN segment intervals from a merged SV VCF."""
    with vcf.VCFReader(vcf_file) as reader, open(out_file, "w") as out_fh:
        callers_from_header = _header_callers(reader.header_str)
        for var in reader:
            svtype = var["INFO"].get("SVTYPE")
            if svtype == "BND":
                continue
            start = int(var["POS"])
            if var["INFO"].get("END"):
                end = int(leading_float(var["INFO"]["END"]))
            else:
                end = start + abs(int(leading_float(var["INFO"].get("SVLEN"))))
            probes = var["INFO"].get("PROBES", 0)
            if var["INFO"].get("FOLD_CHANGE_LOG"):
                fold = var["INFO"]["FOLD_CHANGE_LOG"]
            elif var["INFO"].get("gatkCN"):
                fold = leading_float(var["INFO"]["gatkCN"]) / 2
            elif svtype == "DEL":
                fold = "DEL"
            else:
                fold = "AMP"
            callers = (
                "&".join(callers_from_header)
                if var["INFO"].get("set") == "Intersection"
                else var["INFO"].get("set", "")
            )
            sign = "-" if svtype == "DEL" else "+"
            pr, sr = _manta_pr_sr(var, sample_id) if "manta" in callers else (0, 0)
            print(
                f"{var['CHROM']}\t{start}\t{end}\t{probes}\t{fold}\t{sign}\t{callers}\t{pr}:{sr}",
                file=out_fh,
            )
    return out_file


def coyote_segmentator(
    intersect_bed,
    sample_id,
    panel_files,
    normal=False,
    out_prefix=None,
    panel_out=None,
    raw_out=None,
):
    """Create coyote CN segment BED files from precomputed bedtools intersect output."""
    panel_genes = {}
    for panel in panel_files.split(","):
        _read_panel(panel, panel_genes)

    out_prefix = out_prefix or sample_id
    filtered_path = panel_out or f"{out_prefix}.cn-segments.panel.bed"
    unfiltered_path = raw_out or f"{out_prefix}.cn-segments.bed"
    groups = []
    previous_region = None
    genes = []
    previous_bed = None
    with open(intersect_bed) as in_fh:
        for line in in_fh:
            fields = line.rstrip("\n").split("\t")
            if len(fields) < 8:
                continue
            region = f"{fields[0]}\t{fields[1]}\t{fields[2]}\t"
            bed_str = fields[:8]
            if previous_region and region != previous_region:
                groups.append((previous_bed, genes))
                genes = []
            genes.append(fields[-1])
            previous_region = region
            previous_bed = bed_str
    if previous_bed:
        groups.append((previous_bed, genes))

    with open(filtered_path, "w") as filtered_fh, open(unfiltered_path, "w") as unfiltered_fh:
        for bed_fields, gene_list in groups:
            gene_conc = ",".join(gene_list)
            match_count, interesting = _panel_gene_hits(gene_conc, panel_genes)
            if match_count > 0:
                filtered_fh.write("\t".join(bed_fields[:6]))
                filtered_fh.write(f"\t{gene_conc}\t{interesting}\t")
                filtered_fh.write("\t".join(bed_fields[6:8]))
                if normal:
                    filtered_fh.write("\tNORMAL")
                filtered_fh.write("\n")
            unfiltered_fh.write("\t".join(bed_fields[:6]))
            unfiltered_fh.write(f"\t{gene_conc}\t")
            unfiltered_fh.write("\t".join(bed_fields[6:8]) + "\n")
    return filtered_path, unfiltered_path


def build_parser():
    parser = ArgumentParser(description="Create coyote CN segment BED files from a merged SV VCF.")
    parser.add_argument("--vcf", help="Merged SV VCF.")
    parser.add_argument("--segments-out", help="Write unannotated segment intervals and exit.")
    parser.add_argument("--intersect-bed", help="Input BED produced by bedtools intersect -loj.")
    parser.add_argument("--sample-id", required=True, help="Sample ID.")
    parser.add_argument("--panel", help="Comma-delimited panel definition files.")
    parser.add_argument("--normal", action="store_true", help="Mark filtered segments as NORMAL.")
    parser.add_argument("--out-prefix", help="Output prefix. Defaults to sample ID.")
    parser.add_argument("--panel-out", help="Output BED with panel-matched segments.")
    parser.add_argument("--raw-out", help="Output BED with all segments.")
    return parser

def main(argv=None):
    args = build_parser().parse_args(argv)
    if args.segments_out:
        if not args.vcf:
            raise SystemExit("--vcf is required with --segments-out")
        write_coyote_segment_intervals(args.vcf, args.sample_id, args.segments_out)
        return 0
    if not args.intersect_bed:
        raise SystemExit("--intersect-bed is required unless --segments-out is used")
    if not args.panel:
        raise SystemExit("--panel is required when formatting intersected segments")
    coyote_segmentator(
        args.intersect_bed,
        args.sample_id,
        args.panel,
        args.normal,
        args.out_prefix,
        args.panel_out,
        args.raw_out,
    )
    return 0

if __name__ == "__main__":
    raise SystemExit(main())
